packages/core-contracts/contracts/token/ERC721.sol

```diff
@@ -0,0 +1,232 @@
+        }
+
+        store.name = tokenName;
+        store.symbol = tokenSymbol;
+        store.baseTokenURI = initBaseTokenURI;
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return
+            interfaceId == this.supportsInterface.selector || // ERC165
+            interfaceId == type(IERC721).interfaceId ||
+            interfaceId == type(IERC721Metadata).interfaceId;
+    }
+
+    function name() external view override returns (string memory) {
+        return _erc721Store().name;
+    }
+
+    function symbol() external view override returns (string memory) {
+        return _erc721Store().symbol;
+    }
+
+    function tokenURI(uint256 tokenId) external view override returns (string memory) {
+        if (!_exists(tokenId)) {
+            revert InvalidTokenId(tokenId);
+        }
+
+        string memory baseURI = _erc721Store().baseTokenURI;
+        return bytes(baseURI).length > 0 ? string(abi.encodePacked(baseURI, _toString(tokenId))) : "";
+    }
+
+    function balanceOf(address holder) public view override returns (uint) {
+        if (holder == address(0)) {
+            revert InvalidOwner(holder);
+        }
+        return _erc721Store().balanceOf[holder];
+    }
+
+    function ownerOf(uint256 tokenId) public view override returns (address) {
+        if (!_exists(tokenId)) {
+            revert InvalidTokenId(tokenId);
+        }
+        return _erc721Store().ownerOf[tokenId];
+    }
+
+    function approve(address to, uint256 tokenId) external payable virtual override {
+        ERC721Store storage store = _erc721Store();
+        address holder = store.ownerOf[tokenId];
+
+        if (to == holder) {
+            revert InvalidTo(to);
+        }
+
+        if (msg.sender != holder && !isApprovedForAll(holder, msg.sender)) {
+            revert Unauthorized(msg.sender);
+        }
+
+        store.tokenApprovals[tokenId] = to;
+
+        emit Approval(holder, to, tokenId);
+    }
+
+    function getApproved(uint256 tokenId) external view virtual override returns (address) {
+        if (!_exists(tokenId)) {
+            revert InvalidTokenId(tokenId);
+        }
+
+        return _erc721Store().tokenApprovals[tokenId];
+    }
+
+    function setApprovalForAll(address operator, bool approved) external virtual override {
+        if (msg.sender == operator) {
+            revert InvalidOperator(operator);
+        }
+
+        _erc721Store().operatorApprovals[msg.sender][operator] = approved;
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function isApprovedForAll(address holder, address operator) public view virtual override returns (bool) {
+        return _erc721Store().operatorApprovals[holder][operator];
+    }
+
+    function transferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public payable virtual override {
+        _transfer(from, to, tokenId);
```

Comment:
As in OZ, authorization checks should go closer to the interface, and lower level internal functions should just do what they do, no questions asked. This is a pattern followed by OZ and most Solidity libraries, that we want to maintain.

It might make sense to hoist checks to lower level functions, but as discussed, let's stay more aligned with OZ.

Reply:
Now I got the rationale behind, and with that reasoning makes sense. Fixed